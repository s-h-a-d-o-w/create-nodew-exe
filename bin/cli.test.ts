import crypto from "crypto";
import { readFile, unlink } from "node:fs/promises";
import { MockInstance, vi } from "vitest";

async function mockCli(args: string[], expectedExitCode: number) {
  process.argv = ["node", "./bin/cli.js"].concat(args);

  try {
    await import("./cli.js");
  } catch (error) {
    expect(error).toMatchSnapshot("thrown");
  }

  expect(spyLog.mock.calls).toMatchSnapshot("log");
  expect(spyError.mock.calls).toMatchSnapshot("error");
  // eslint-disable-next-line @typescript-eslint/unbound-method
  expect(process.exit).toHaveBeenCalledWith(expectedExitCode);
}

let spyLog: MockInstance, spyError: MockInstance;
beforeAll(() => {
  vi.spyOn(process, "exit");

  // mockImplementation() to suppress all that CLI console output in the test logs
  // Output will still be visible in the snapshots because of .mock.calls
  spyLog = vi.spyOn(console, "log").mockImplementation(() => {});
  spyError = vi.spyOn(console, "error").mockImplementation(() => {});
});

beforeEach(() => {
  // Without resetting, yargs would remember args from previous tests
  vi.resetModules();

  // Purge previous console output before each test
  vi.clearAllMocks();
});

test("--help", async () => {
  await mockCli(["--help"], 0);
});

test("no arguments", async () => {
  await mockCli([], 1);
});

test("only src, no dst", async () => {
  await mockCli(["src"], 1);
});

test("correct conversion", async () => {
  const regularExecutable = "./bin/testData/node.exe";
  const silentExecutable = "./bin/testData/nodew.exe";
  process.argv = ["node", "./bin/cli.js", regularExecutable, silentExecutable];

  await import("./cli.js");

  // Unless the source .exe changes, these check sums should obviously always stay the same
  expect(
    crypto
      .createHash("md5")
      .update(await readFile(regularExecutable))
      .digest("hex"),
  ).toEqual("037c3a6fcd6719e26d1bfc7a1a835a67");
  expect(
    crypto
      .createHash("md5")
      .update(await readFile(silentExecutable))
      .digest("hex"),
  ).toEqual("d1bf9c7e48771ed322827e5549716484");

  expect(spyLog.mock.calls).toMatchSnapshot("log");
  expect(spyError.mock.calls).toMatchSnapshot("error");

  await unlink(silentExecutable);
});
