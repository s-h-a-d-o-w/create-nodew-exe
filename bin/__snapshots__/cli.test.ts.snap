// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`--help > error 1`] = `[]`;

exports[`--help > log 1`] = `
[
  [
    "Converts node.exe into a silent executable (no terminal will pop up) by changing
 only one byte.

Usage: cli.js [options] <src> <dst>

Options:
      --version  Show version number                                   [boolean]
  -h, --help     Show help                                             [boolean]",
  ],
]
`;

exports[`--help > thrown 1`] = `[Error: process.exit unexpectedly called with "0"]`;

exports[`correct conversion > error 1`] = `[]`;

exports[`correct conversion > log 1`] = `
[
  [
    "Generating: ",
    "./bin/testData/nodew.exe",
  ],
  [
    "SUCCESS",
  ],
]
`;

exports[`no arguments > error 1`] = `
[
  [
    "You need to specify <src> and <dst> (or use --help for more)",
  ],
]
`;

exports[`no arguments > log 1`] = `[]`;

exports[`no arguments > thrown 1`] = `[Error: process.exit unexpectedly called with "1"]`;

exports[`only src, no dst > error 1`] = `
[
  [
    "You need to specify <src> and <dst> (or use --help for more)",
  ],
]
`;

exports[`only src, no dst > log 1`] = `[]`;

exports[`only src, no dst > thrown 1`] = `[Error: process.exit unexpectedly called with "1"]`;
